man/man1/lrsync.1
.\" Sorry for my English 
.\" --Dmitry Yu Okunev <dyokunev@ut.mephi.ru> 0x8E30679C
.\"
.TH CLSYNC 1 "JUNE 2014" Linux "User Manuals"
.SH NAME
lrsync \- rsync\-like clsync\-wrapper for live syncing
.SH SYNOPSIS
.B lrsync [ lrsync, clsync and rsync options ]
.SH DESCRIPTION
.B lrsync
executes
.BR clsync "(1)"
with appropriate arguments.

.B lrsync
executes "clsync \-\-help" to find out long options of clsync
and passes such options to it. If option is not for clsync it passes
it to rsync via
.I sync\-handler\-arguments
of clsync [see 
.BR clsync "(1)]."

The
.B lrsync
executes
.B clsync
using the next template:
.RS
.I path\-to\-clsync
\-Klrsync \-Mrsyncdirect \-S
.I path\-to\-rsync
\-W
.I source\-directory
\-D
.I destination\-directory
\-L
.I lists\-dir\-path
.I clsync\-arguments
\-\- %RSYNC\-ARGS%
.I rsync\-arguments
.RE

Substitutions:
.RS
.B path\-to\-clsync
.RS
Equals to ENV "CLSYNC_PATH" if it is set. Otherwise it equals to "clsync".
.RE
.B path\-to\-rsync
.RS
Equals to ENV "RSYNC_PATH" if it is set. Otherwise it equals to "rsync".
.RE
.B source\-directory
.RS
The first non\-option lrsync argument.
.RE
.B destination\-directory
.RS
The second non\-option lrsync argument.
.RE
.B lists\-dir\-path
.RS
Equals to ENV "CLSYNC_LISTS_PATH" if it is set. Otherwise it equals to
"/tmp/.lrsync".
.RE
.B clsync\-arguments
.RS
Equals to parsed clsync long options from the lrsync arguments.
.RE
.B rsync\-arguments
.RS
The rest part of lrsync arguments.
.RE
.RE

.SH OPTIONS

Too see
.B rsync
options use "man 1 rsync".

Too see
.B clsync
options use "man 1 clsync".

.B "lrsync options:"
.RS
.B \-\-help
.RS
Prints options lists of lrsync, clsync and rsync.
.RE
.B \-\-version
.RS
Prints version of lrsync, clsync and rsync.
.RE
.B \-\-clsync\-command\-only
.RS
Prints the clsync command to be runned but doesn't run it.
.RE
.RE

.SH AUTHOR
Dmitry Yu Okunev <dyokunev@ut.mephi.ru> 0x8E30679C

.SH SUPPORT
You can get support on official IRC-channel of clsync in Freenode "#clsync"
or on github's issue tracking system of repository
"https://github.com/xaionaro/lrsync".

.SH SEE ALSO
.BR clsync "(1),"
.BR rsync "(1),"
.BR inotify "(7)"


lrsync.c
/*
    lrsync - rsync-like clsync wrapper

    Copyright (C) 2014  Dmitry Yu Okunev <dyokunev@ut.mephi.ru> 0x8E30679C

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <unistd.h>	// fork()
#include <stdlib.h>	// getenv()
#include <stdarg.h>	// va_start()
#include <string.h>	// memset()
#include <errno.h>	// errno
#include <sys/types.h>	// waitpid()
#include <sys/wait.h>	// waitpid()

#include "common.h"
#include "error.h"
#include "malloc.h"
#include "lrsync.h"

int forkexecwaitvp(const char *const bin_path, const char **const argv)
{
	pid_t pid;
	int status;

	// Forking
	pid = fork();
	switch(pid) {
		case -1: 
			error("Cannot fork().");
			return errno;
		case  0:
			argv[0] = bin_path;
			execvp(bin_path, (char *const *)argv);
			return errno;
	}

	if(waitpid(pid, &status, 0) != pid) {
		error("Cannot waitid().");
		return errno;
	}

	// Return
	int exitcode = WEXITSTATUS(status);

	return exitcode;
}

#define arglist2argv(argv, firstarg, COPYARG) {\
	va_list arglist;\
	va_start(arglist, firstarg);\
\
	int i = 1;\
	do {\
		char *arg;\
		if(i >= MAXARGUMENTS) {\
			error("Too many arguments (%i >= %i).", i, MAXARGUMENTS);\
			return ENOMEM;\
		}\
		arg = (char *)va_arg(arglist, const char *const);\
		argv[i] = arg!=NULL ? COPYARG : NULL;\
	} while(argv[i++] != NULL);\
	va_end(arglist);\
}

const char *path_rsync() {
	char *rsync_path;
	rsync_path = getenv("RSYNC_PATH");
	if (rsync_path == NULL)
		rsync_path = DEFAULT_RSYNC_PATH;

	return rsync_path;
}

const char *path_clsync() {
	char *clsync_path;
	clsync_path = getenv("CLSYNC_PATH");
	if (clsync_path == NULL)
		clsync_path = DEFAULT_CLSYNC_PATH;

	return clsync_path;
}

const char *path_listsdir() {
	char *listsdir_path;
	listsdir_path = getenv("CLSYNC_LISTS_PATH");
	if (listsdir_path == NULL)
		listsdir_path = DEFAULT_LISTSDIR_PATH;

	return listsdir_path;
}

static inline int exec_rsync_argv(const char **argv)
{
	return forkexecwaitvp(path_rsync(), argv);
}

static inline int exec_clsync_argv(const char **argv)
{
	return forkexecwaitvp(path_clsync(), argv);
}

int exec_rsync(ctx_t *const ctx_p, ...)
{
	char **argv = (char **)xcalloc(sizeof(char *), MAXARGUMENTS);
	memset(argv, 0, sizeof(char *)*MAXARGUMENTS);
	arglist2argv(argv, ctx_p, arg);

	int rc = exec_rsync_argv((const char **)argv);
	free(argv);
	return rc;
}

int exec_clsync(ctx_t *const ctx_p, ...)
{
	char **argv = (char **)xcalloc(sizeof(char *), MAXARGUMENTS);
	memset(argv, 0, sizeof(char *)*MAXARGUMENTS);
	arglist2argv(argv, ctx_p, arg);

	int rc = exec_clsync_argv((const char **)argv);
	free(argv);
	return rc;
}

int print_cmd(const char *const *const argv) {
	const char *const *argv_p;

	argv_p = argv;
	if (*argv_p != NULL)
		printf("%s ", *(argv_p++));

	while (*argv_p != NULL)
		printf("\\\n  %s ", *(argv_p++));

	printf("\n\n");
	return 0;
}

static inline void push_arg(const char **argv, int *count_p, const char *arg) {
	if ((*count_p) >= MAXARGUMENTS) {
		errno = E2BIG;
		critical("Too many arguments");
	}
	argv[(*count_p)++] = arg;
	return;
}

int lrsync(ctx_t *const ctx_p)
{
	int          count = 0;
	const char  *argv[MAXARGUMENTS];
	char       **argv_p;

	if (ctx_p->dir_from == NULL)
		errno=EINVAL, critical("Source is not entered");
	if (ctx_p->dir_to   == NULL)
		errno=EINVAL, critical("Destination is not entered");

	push_arg(argv, &count, path_clsync());
	push_arg(argv, &count, "-Klrsync");
	push_arg(argv, &count, "-Mrsyncdirect");
	push_arg(argv, &count, "-S");
	push_arg(argv, &count, path_rsync());
	push_arg(argv, &count, "-W");
	push_arg(argv, &count, ctx_p->dir_from);
	push_arg(argv, &count, "-D");
	push_arg(argv, &count, ctx_p->dir_to);
	push_arg(argv, &count, "-L");
	push_arg(argv, &count, path_listsdir());

	argv_p = ctx_p->clsync_argv;
	while (*argv_p != NULL)
		push_arg(argv, &count, *(argv_p++));

	push_arg(argv, &count, "--");
	push_arg(argv, &count, "%RSYNC-ARGS%");

	argv_p = ctx_p->rsync_argv;
	while (*argv_p != NULL)
		push_arg(argv, &count, *(argv_p++));

	if (ctx_p->flags[FL_COMMANDONLY])
		return print_cmd(argv);

	return exec_clsync_argv(argv);
}
